chrome-extension/popup.js

```diff
@@ -127,37 +129,125 @@ function downloadButtonCallback(chapter_list_url) {
+  var match2 = re2.exec(url);
+
+  if(match != null && match2 != null)
+    return 2;     // return 2 if the page has loaded 100 books in the Safari Queue page
+  else if (match != null)
+    return 1;     // return 1 if the page is in Safari Queue page
+  else 
+    return 0;     // return 0 if the page is NOT in Safari Queue page
+}
+
+function clickedEvent(url)
+{
+  var re = new RegExp('(\.+://www.safaribooksonline.com/library/view/\.+/\\d+/)\.*');
+  var match = re.exec(url);
+
+  if (match != null) {
+    book_url = match[1] + "#toc";
+    downloadButtonCallback(book_url);
+  }
+  else {
+    renderStatus("Unknown url, not able to get book chapter url");
+  }
+}
+
 document.addEventListener('DOMContentLoaded', function() {
   var button = document.getElementById('download-button');
+  var loadAllInQueueButton = document.getElementById('load-all-in-queue-button');
   button.disabled = true;
+  loadAllInQueueButton.disabled = false;
 
   getCurrentTab(function(tab) {
-    var book_name = getBookNameFromUrl(tab.url);
-    if (book_name == null) {
-        setBookName("Unable to get book name, please change another url and try again")
-        return;
-    }
-    else {
-        setBookName(book_name);
-    }
+    var book_name;
 
-    // Inject content script and button callback
-    chrome.tabs.executeScript(null, {file: "content.js"});
+    chrome.tabs.onUpdated.addListener(function(tabId , info) {
+      if(tabId == tab.id && info.status == "complete"){
+        location.reload();
+        window.close();
+      }
+    });
 
-    var button = document.getElementById('download-button');
-    button.disabled = false;
-    button.addEventListener('click', function() {
-      tab_url = tab.url;
-      var re = new RegExp('(\.+://www.safaribooksonline.com/library/view/\.+/\\d+/)\.*');
-      var match = re.exec(tab_url)
-      if (match != null) {
-        book_url = match[1] + "#toc"
-        downloadButtonCallback(book_url);
+    loadAllInQueueButton.addEventListener('click', function() {
+      
+        chrome.tabs.executeScript( null, {code: "location.href='https://www.safaribooksonline.com/s/?limit=100'; "}, 
+          function() {
+            loadAllInQueueButton.textContent = "Loading...";
+            loadAllInQueueButton.disabled = true;
+          });
+      
+
+    });
+
+    if (isInQuenePage(tab.url))
+    {
+      loadAllInQueueButton.disabled = isInQuenePage(tab.url) == 2 ? true : false;
+
+      var urls = [];
+      chrome.tabs.executeScript( null, {
+        code: "var b = []; a = document.getElementsByClassName('next title-block js-bit-title t-bit-link'); for (var i = 0 ; i < a.length ; i++) b[i] = 'https://www.safaribooksonline.com' + a[i].getAttribute('href'); b"
+      }, function(results){ 
+
+        urls = results[0];
+        setBookName("Warning! If you click the 'Download' button in this page, all books in the Safari Quene will be downloaded.\n\n" + 
```

Comment:
"download all books concurently" doesn't sounds like a good idea, espcially when you allow at most 100 books to download at same time. Chrome will fork more than 100 process to download these books and certainly crash user's computer. (You can test this out) Let's stick to one book at a time or set a concurrent upper limit.